src/hotspot/share/logging/logDecorations.cpp

```diff
@@ -53,89 +48,93 @@ const char* LogDecorations::host_name() {
-    position = create_##full_name##_decoration(position) + 1; \
-  } else { \
-    _decoration_offset[LogDecorators::full_name##_decorator] = NULL; \
-  }
+LogDecorations::LogDecorations(LogLevelType level, const LogTagSet &tagset, const LogDecorators &decorators) :
+  // When constructing the LogDecorations we resolve values for the requested decorators.
+  //
+  // _millis: needed for "time", "utctime", "timemillis":
+  _millis(
+      (decorators.is_decorator(LogDecorators::time_decorator) ||
+       decorators.is_decorator(LogDecorators::utctime_decorator) ||
+       decorators.is_decorator(LogDecorators::timemillis_decorator)) ? os::javaTimeMillis() : 0),
+  // _nanos: needed for "timenanos"
+  _nanos(decorators.is_decorator(LogDecorators::timenanos_decorator) ? os::javaTimeNanos() : 0),
+  // _elapsed_seconds: needed for "uptime", "uptimemillis", "uptimenanos"
+  _elapsed_seconds(
+      (decorators.is_decorator(LogDecorators::uptime_decorator) ||
+       decorators.is_decorator(LogDecorators::uptimemillis_decorator) ||
+       decorators.is_decorator(LogDecorators::uptimenanos_decorator)) ? os::elapsedTime() : 0),
+  // pid
+  _pid(decorators.is_decorator(LogDecorators::pid_decorator) ? os::current_process_id() : 0),
+  // tid
+  _tid(decorators.is_decorator(LogDecorators::tid_decorator) ? os::current_thread_id() : 0),
+  // the rest is handed down by the caller
+  _level(level), _tagset(tagset)
+{
+}
+
+void LogDecorations::print_decoration(LogDecorators::Decorator decorator, outputStream* st) const {
+  switch(decorator) {
+#define DECORATOR(name, abbr) case LogDecorators:: name##_decorator: print_##name##_decoration(st); break;
   DECORATOR_LIST
 #undef DECORATOR
+    default: ShouldNotReachHere();
+  }
 }
 
-#define ASSERT_AND_RETURN(written, pos) \
-    assert(written >= 0, "Decorations buffer overflow"); \
-    return pos + written;
-
-char* LogDecorations::create_time_decoration(char* pos) {
-  char* buf = os::iso8601_time(pos, 29);
-  int written = buf == NULL ? -1 : 29;
-  ASSERT_AND_RETURN(written, pos)
+const char* LogDecorations::decoration(LogDecorators::Decorator decorator, char* buf, size_t buflen) const {
+  stringStream ss(buf, buflen);
+  print_decoration(decorator, &ss);
+  return buf;
 }
 
-char* LogDecorations::create_utctime_decoration(char* pos) {
-  char* buf = os::iso8601_time(pos, 29, true);
-  int written = buf == NULL ? -1 : 29;
-  ASSERT_AND_RETURN(written, pos)
+void LogDecorations::print_time_decoration(outputStream* st) const {
+  char buf[29];
+  st->print_raw(os::iso8601_time(_millis, buf, sizeof(buf), false));
 }
 
-char * LogDecorations::create_uptime_decoration(char* pos) {
-  int written = jio_snprintf(pos, DecorationsBufferSize - (pos - _decorations_buffer), "%.3fs", os::elapsedTime());
-  ASSERT_AND_RETURN(written, pos)
+void LogDecorations::print_utctime_decoration(outputStream* st) const {
+  char buf[29];
+  st->print_raw(os::iso8601_time(_millis, buf, sizeof(buf), true));
```

Comment:
Same as for `LogDecorations::print_time_decoration()`.

Reply:
+1